contrib/src/compression.rs

```diff
@@ -0,0 +1,129 @@
+//! This module provides brotli ang gzip compression for all non-image
+//! responses for requests that send Accept-Encoding br and gzip. If
+//! accepted, brotli compression is preferred over gzip.
+//!
+//! To add this feature to your Rocket application, use
+//! .attach(rocket_contrib::Compression::fairing())
+//! to your Rocket instance. Note that you must add the
+//! "compression" feature for brotli and gzip compression to your rocket_contrib
+//! dependency in Cargo.toml. Additionally, you can load only brotli compression
+//! using "brotli_compression" feature or load only gzip compression using
+//! "gzip_compression" in your rocket_contrib dependency in Cargo.toml.
+//!
+//! In the brotli algorithm, quality is set to 2 in order to have really fast
+//! compressions with compression ratio similar to gzip. Also, text and font
+//! compression mode is set regarding the Content-Type of the response.
+//!
+//! In the gzip algorithm, quality is the default (9) in order to have good
+//! compression ratio.
+//!
+//! For brotli compression, the rust-brotli crate is used.
+//! For gzip compression, flate2 crate is used.
+
+use std::io::Cursor;
+#[cfg(feature = "gzip_compression")]
+use std::io::Write;
+
+use rocket::{Request, Response};
+use rocket::http::Header;
+use rocket::fairing::{Fairing, Info, Kind};
+
+#[cfg(feature = "brotli_compression")]
+use brotli;
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2;
+#[cfg(feature = "gzip_compression")]
+use flate2::write::GzEncoder;
+
+#[derive(Debug, Default)]
+pub struct Compression;
+
+impl Compression {
+    /// This function creates a Compression to be used in your Rocket
+    /// instance. Add ```.attach(rocket_contrib::Compression::fairing())```
+    /// to your Rocket instance to use this fairing.
+    ///
+    /// # Returns
+    ///
+    /// A Compression instance.
+    pub fn fairing() -> Compression {
+        Compression::default()
+    }
+}
+
+impl Fairing for Compression {
+    fn info(&self) -> Info {
+        Info {
+            name: "Brotli and gzip compressors for responses",
+            kind: Kind::Response,
+        }
+    }
+
+    fn on_response(&self, request: &Request, response: &mut Response) {
+        let content_type = response.content_type();
+        // Images must not be compressed
+        if let Some(ref content_type) = content_type {
+            if content_type.top() == "image" {
+                return;
+            }
+        }
+
+        // The compression is done if the request supports brotli or gzip and
+        // the corresponding feature is enabled
+        if cfg!(feature = "brotli_compression")
+            && request
+                .headers()
+                .get("Accept-Encoding")
+                .flat_map(|e| e.trim().split(","))
```

Comment:
I use trim in case any browser set the header with spaces (for instance: "Accept-Encoding: deflate, gzip, br").

In Firefox and Chrome this does not happen, but I think that this makes the fairing more robust.